### PR TITLE
docs(components): Disclosure Component Proposal

### PR DESCRIPTION
**Pull request by ryanjones**
### April 2021 update:
@kingstonfung & friends has now taken over this PR and will be responsible for carrying this forward

## Motivations

Add a disclosure component to Atlantis

### Added

- Disclosure component proposal document

### Mockup
image

---

[In Atlantis we use Github's built in pull request reviews](https://help.github.com/en/articles/about-pull-request-reviews).

Random photo of Atlantis

### CODE REVIEW

**Review by seanhealy — CHANGES_REQUESTED** [review]
From a programmatic point of view this is looking pretty good. Had one suggestion about props for it to run in controlled or uncontrolled modes.

**Comment by rebecca-li** [comment]
Quick example [here](https://www.figma.com/proto/NSFGqkxgycP8JbEwmaDW8d/Design-System-Contribution-Disclosure?node-id=131%3A10&viewport=4200%2C-1004%2C1.7579501867294312&scaling=min-zoom). We can revisit the Figma file and add more details about usage and specs later (e.g. when to use disclosure vs. accordion)

**Comment by kingstonfung** [comment]
I'll be taking over this draft and try to push forward for a `<Disclosure>` component. Updated proposal incoming

**Comment by kingstonfung** [comment]
🎉  reviving this PR and sent in for review 🎉

File: docs/CONTRIBUTING.md [hunk 1]
```diff
@@ -40,7 +40,7 @@ as a guide for writing proposals.
 2. Copy the proposal template and place it
    under `docs/proposals/<component_name>/README.md`
 3. Fill out the proposal template
-4. Submit a PR for the propsoal
+4. Submit a PR for the proposal
```
Comment (chris-at-jobber) [inline docs/CONTRIBUTING.md:43]: You can drop this typo fix if you want, I have a whole PR up changing the contributing guidelines 😂

File: docs/proposals/Disclosure/README.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
+# Disclosure
+
+A `<Disclosure>` component will reveal and hide additional information that our
+users may care about, should they choose to learn more about it.
+
+## Design Usage and Guidelines
+
+- use when you want to reduce feelings of being overwhelmed and to reduce
+  distractions
+- used for lightweight editing, such as selecting a value for a setting
```
Comment (chris-at-jobber) [inline docs/proposals/Disclosure/README.md:10]: I'm a little fuzzy on what we mean by this line - are we trying to convey something like "can be used to hide optional settings controls"?

File: docs/proposals/Disclosure/README.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+## Mockup
+
+!["Disclosure Mockup"](https://gist.githubusercontent.com/kingstonfung/b882aa211f4213b1ac89bd515dc48a50/raw/efef596279fa0070d26b67cd364e533ca3d572bd/image2.png)
+
+## Interface
+
+```jsx
+<Disclosure title="Advanced Options">
+  <Content>Example content here</Content>
```
Comment (chris-at-jobber) [inline docs/proposals/Disclosure/README.md:72]: One thing we could clarify here or at the actual code review level is whether we want to have a [`<Content>` component](https://atlantis.getjobber.com/components/content) in here to provide spacing of the elements by default, or if <Content> in this example just means "children"

File: docs/proposals/Disclosure/README.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+| defaultOpen     | boolean      | false     | This sets the default open state of the disclosure. By default the disclosure is closed. For use when the component is being used as an [Uncontrolled](https://reactjs.org/docs/uncontrolled-components.html) component |
+| onRequestToggle | (() => void) | void      | This function would toggle the open state. For use when the component is being used as a [Controlled](https://reactjs.org/docs/forms.html#controlled-components) component                                              |
+
+## Notes
+
+**Trade-offs, considerations, and things we want to remember**
+
+- `onchange` event for tracking when opens
+- how it’ll behave and look when it’s stacking
+- `<Heading>` on the title is open for other variations in the future
```
Comment (chris-at-jobber) [inline docs/proposals/Disclosure/README.md:91]:
```suggestion
- Assumption: `<Heading level=4>` should solve most use cases for this component. We did consider the possibility of opening this up to other levels (3, maybe?) but for a simpler end-user experience are not making the heading level configurable from the component API at this time.
```

File: docs/proposals/Disclosure/README.md [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+| open            | boolean      | undefined | Control if the disclosure component is open or closed                                                                                                                                                                   |
+| defaultOpen     | boolean      | false     | This sets the default open state of the disclosure. By default the disclosure is closed. For use when the component is being used as an [Uncontrolled](https://reactjs.org/docs/uncontrolled-components.html) component |
+| onRequestToggle | (() => void) | void      | This function would toggle the open state. For use when the component is being used as a [Controlled](https://reactjs.org/docs/forms.html#controlled-components) component                                              |
+
+## Notes
+
+**Trade-offs, considerations, and things we want to remember**
+
+- `onchange` event for tracking when opens
+- how it’ll behave and look when it’s stacking
```
Comment (chris-at-jobber) [inline docs/proposals/Disclosure/README.md:90]:
```suggestion
- While this is roughly based on a legacy Jobber implementation of an accordion, we haven't fleshed out exactly how Disclosure will behave and look if used in an accordion fashion
```

File: docs/proposals/Disclosure/README.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+| title           | string       | -         | Title for the disclosure pane                                                                                                                                                                                           |
+| open            | boolean      | undefined | Control if the disclosure component is open or closed                                                                                                                                                                   |
+| defaultOpen     | boolean      | false     | This sets the default open state of the disclosure. By default the disclosure is closed. For use when the component is being used as an [Uncontrolled](https://reactjs.org/docs/uncontrolled-components.html) component |
+| onRequestToggle | (() => void) | void      | This function would toggle the open state. For use when the component is being used as a [Controlled](https://reactjs.org/docs/forms.html#controlled-components) component                                              |
+
+## Notes
+
+**Trade-offs, considerations, and things we want to remember**
+
+- `onchange` event for tracking when opens
```
Comment (chris-at-jobber) [inline docs/proposals/Disclosure/README.md:89]: @kingstonfung / @seanhealy do you recall what this point is referring to?

File: docs/proposals/Disclosure/README.md [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+- how it’ll behave and look when it’s stacking
+- `<Heading>` on the title is open for other variations in the future
+
+**Places in Jobber’s codebase something similar exists (don’t include in
```
Comment (chris-at-jobber) [inline docs/proposals/Disclosure/README.md:93]: These examples should be removed (I'll clarify the verbiage in the Figma template)

File: docs/proposals/Disclosure/README.md [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+- **Title** should be informative and label the type of content grouped in the
+  body content in it
+  - use `<Heading level={4}>` component
+- **Content** should be actionable and clear. May contain:
+  - good ol’ plain text
+  - accepts any React component as a child
+  - except: `<Page>`, `<Table>`
+
+## Accessibility
```
Comment (chris-at-jobber) [inline docs/proposals/Disclosure/README.md:39]: ❤️

**Review by chris-at-jobber — CHANGES_REQUESTED** [review]
Just a few clarification things docs-wise!

File: docs/proposals/Disclosure/README.md (hunk 2 above)
Comment (kingstonfung) [reply docs/proposals/Disclosure/README.md:10]: I agree, that sounds much better

File: docs/proposals/Disclosure/README.md (hunk 3 above)
Comment (kingstonfung) [reply docs/proposals/Disclosure/README.md:72]: I think we can just do `Example content here` without the `<Content>` for clarity. We intended to meant just "children" and not necessary have to be `<Content>` wrapped.

File: docs/proposals/Disclosure/README.md (hunk 6 above)
Comment (kingstonfung) [reply docs/proposals/Disclosure/README.md:87]: @chris-at-jobber : The mention here I believe was to set out a reminder that we _may_ , in the future, setup a handler that receives expanded/collapsed events. For example so other parts of a page, SPA or not, can subscribe to this and react accordingly. Hope this makes sense. Thanks!

File: docs/CONTRIBUTING.md (hunk 1 above)
Comment (kingstonfung) [reply docs/CONTRIBUTING.md]: Dropped the changes against this file. Thanks for mentioning this!

**Comment by kingstonfung** [comment]
@chris-at-jobber : Addressed items listed out from previous comments, PLMK if you spot anything else in there. Thank you!

File: docs/proposals/Disclosure/README.md (hunk 6 above)
Comment (chris-at-jobber) [reply docs/proposals/Disclosure/README.md:87]: Makes sense! Let's include that detail here:

```suggestion
- We may, in the future, setup an `onchange` handler that receives expanded/collapsed events. For example so other parts of a page, SPA or not, can subscribe to this and react accordingly.
```

**Review by chris-at-jobber — APPROVED** [review]
I'm good with this! I'd like to add that bit of detail to the future `onchange` event but once that's in let's get to building this thing 🎉

**Comment by kingstonfung** [comment]
Thanks for the approval!🥳  

`59a79f2` is the commit (after the approval) is just the updated sentence elaborating `onchange` considerations, as per https://github.com/GetJobber/atlantis/pull/351#discussion_r610321015 directly above

**Pull request merged**